File: ring-android/app/src/main/java/cx/ring/account/RenameDeviceDialog.java
/*
 *  Copyright (C) 2004-2021 Savoir-faire Linux Inc.
 *
 *  Author: Adrien Beraud <adrien.beraud@savoirfairelinux.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package cx.ring.account;

import android.app.Dialog;
import android.os.Bundle;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;

import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;

import androidx.fragment.app.DialogFragment;
import cx.ring.R;
import cx.ring.databinding.DialogDeviceRenameBinding;

public class RenameDeviceDialog extends DialogFragment {
    public static final String DEVICENAME_KEY = "devicename_key";
    static final String TAG = RenameDeviceDialog.class.getSimpleName();
    private RenameDeviceListener mListener = null;
    private DialogDeviceRenameBinding binding;

    public void setListener(RenameDeviceListener listener) {
        mListener = listener;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        binding = DialogDeviceRenameBinding.inflate(getActivity().getLayoutInflater());

        binding.ringDeviceNameTxt.setText(getArguments().getString(DEVICENAME_KEY));
        binding.ringDeviceNameTxt.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                boolean validationResult = validate();
                if (validationResult) {
                    getDialog().dismiss();
                }
                return validationResult;
            }
            return false;
        });

        final AlertDialog dialog = new MaterialAlertDialogBuilder(requireContext())
                .setView(binding.getRoot())
                .setTitle(R.string.rename_device_title)
                .setMessage(R.string.rename_device_message)
                .setPositiveButton(R.string.rename_device_button, null)
                .setNegativeButton(android.R.string.cancel, null)
                .create();
        dialog.setOnShowListener(dialog1 -> {
            Button button = ((AlertDialog) dialog1).getButton(AlertDialog.BUTTON_POSITIVE);
            button.setOnClickListener(view1 -> {
                if (validate()) {
                    dialog1.dismiss();
                }
            });
        });
        dialog.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE);
        return dialog;
    }

    @Override
    public void onDestroy() {
        mListener = null;
        super.onDestroy();
    }

    private boolean checkInput(String input) {
        if (input.isEmpty()) {
            binding.ringDeviceNameTxtBox.setErrorEnabled(true);
            binding.ringDeviceNameTxtBox.setError(getText(R.string.account_device_name_empty));
            return false;
        } else {
            binding.ringDeviceNameTxtBox.setErrorEnabled(false);
            binding.ringDeviceNameTxtBox.setError(null);
        }
        return true;
    }

    private boolean validate() {
        String input = binding.ringDeviceNameTxt.getText().toString().trim();
        if (checkInput(input) && mListener != null) {
            mListener.onDeviceRename(input);
            return true;
        }
        return false;
    }

    public interface RenameDeviceListener {
        void onDeviceRename(String newName);
    }
}

File: jami-android/app/src/main/java/cx/ring/account/RenameDeviceDialog.kt
/*
 *  Copyright (C) 2004-2023 Savoir-faire Linux Inc.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package cx.ring.account

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.KeyEvent
import android.view.WindowManager
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import cx.ring.R
import cx.ring.databinding.DialogDeviceRenameBinding

class RenameDeviceDialog : DialogFragment() {
    private var mListener: RenameDeviceListener? = null
    private var binding: DialogDeviceRenameBinding? = null
    fun setListener(listener: RenameDeviceListener?) {
        mListener = listener
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        binding = DialogDeviceRenameBinding.inflate(layoutInflater)
        binding!!.ringDeviceNameTxt.setText(requireArguments().getString(DEVICENAME_KEY))
        binding!!.ringDeviceNameTxt.setOnEditorActionListener { v: TextView?, actionId: Int, event: KeyEvent? ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                val validationResult = validate()
                if (validationResult) {
                    requireDialog().dismiss()
                }
                return@setOnEditorActionListener validationResult
            }
            false
        }
        return MaterialAlertDialogBuilder(requireContext())
            .setView(binding!!.root)
            .setTitle(R.string.rename_device_title)
            .setMessage(R.string.rename_device_message)
            .setPositiveButton(R.string.rename_device_button, null)
            .setNegativeButton(android.R.string.cancel, null)
            .create()
            .apply {
                setOnShowListener { d: DialogInterface ->
                    val button = (d as AlertDialog).getButton(AlertDialog.BUTTON_POSITIVE)
                    button.setOnClickListener {
                        if (validate()) {
                            d.dismiss()
                        }
                    }
                }
                window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE)
            }
    }

    override fun onDestroy() {
        mListener = null
        super.onDestroy()
    }

    private fun checkInput(input: String): Boolean {
        if (input.isEmpty()) {
            binding?.apply {
                ringDeviceNameTxtBox.isErrorEnabled = true
                ringDeviceNameTxtBox.error = getText(R.string.account_device_name_empty)
            }
            return false
        } else {
            binding?.apply {
                ringDeviceNameTxtBox.isErrorEnabled = false
                ringDeviceNameTxtBox.error = null
            }
        }
        return true
    }

    private fun validate(): Boolean {
        val input = binding!!.ringDeviceNameTxt.text.toString().trim { it <= ' ' }
        if (checkInput(input) && mListener != null) {
            mListener!!.onDeviceRename(input)
            return true
        }
        return false
    }

    interface RenameDeviceListener {
        fun onDeviceRename(newName: String)
    }

    companion object {
        const val DEVICENAME_KEY = "devicename_key"
    }
}